internal/cephfs/groupcontrollerserver.go

```diff
@@ -61,3 +72,563 @@ func (cs *ControllerServer) validateVolumeGroupSnapshotRequest(
+			releasing = true
+
+			continue
+		}
+	}
+
+	if releasing {
+		return cerrors.ErrInProgress
+	}
+
+	return nil
+}
+
+func subvolumeQuiesceWithExpireTimeout(ctx context.Context,
+	requestName string,
+	fsMap map[string]core.SubVolumeQuiesceClient,
+) error {
+	var err error
+
+	var data *admin.QuiesceInfo
+	inProgress := false
+	for _, fm := range fsMap {
+		// reinitialize the expiry timer for the quiesce
+		data, err = fm.SubvolumeQuiesceWithExpireTimeout(ctx, requestName)
+		if err != nil && !errors.Is(err, rados.ErrPermissionDenied) {
+			log.ErrorLog(ctx, "failed to quiesce subvolume with timeout: %v", err)
+
+			return err
+		}
+		state := core.GetQuiesceState(data)
+		if errors.Is(err, rados.ErrPermissionDenied) || state == core.Quiescing {
+			inProgress = true
+
+			continue
+		}
+
+		if state != core.Quiesced {
+			return fmt.Errorf("quiesce operation is in %s state", state)
+		}
+	}
+
+	if inProgress {
+		return cerrors.ErrInProgress
+	}
+
+	return nil
+}
+
+func (cs *ControllerServer) createSnapshotAndAddMapping(
+	ctx context.Context,
+	req *csi.CreateSnapshotRequest,
+	vgo *store.VolumeGroupOptions,
+	vgs *store.VolumeGroupSnapshotIdentifier,
+) (*csi.CreateSnapshotResponse, error) {
+	// Create the snapshot
+	resp, err := cs.CreateSnapshot(ctx, req)
+	if err != nil {
+		// Handle cleanup
+		log.ErrorLog(ctx, "failed to create snapshot: %v", err)
+
+		return nil, err
+	}
+	// Add the snapshot to the volume group journal
+	err = store.VolumeGroupJournal.AddVolumeSnapshotMapping(ctx,
+		vgo.MetadataPool,
+		vgs.ReservedID,
+		req.SourceVolumeId,
+		resp.GetSnapshot().GetSnapshotId())
+	if err != nil {
+		log.ErrorLog(ctx, "failed to add volume snapshot mapping: %v", err)
+		// Delete the last created snapshot as its still not added to the
+		// journal
+		delReq := &csi.DeleteSnapshotRequest{
+			SnapshotId: resp.GetSnapshot().GetSnapshotId(),
+			Secrets:    req.Secrets,
+		}
+		_, dErr := cs.DeleteSnapshot(ctx, delReq)
+		if dErr != nil {
+			log.ErrorLog(ctx, "failed to delete snapshot %s: %v", resp.GetSnapshot().GetSnapshotId(), dErr)
+		}
+
+		return nil, err
+	}
+
+	return resp, nil
+}
+
+// checkIfFSNeedQuiesceRelease checks that do we have snapshots for all the
+// volumes stored in the omap so that we can release the quiesce.
+func checkIfFSNeedQuiesceRelease(vgs *store.VolumeGroupSnapshotIdentifier, volIDs []string) bool {
+	if vgs == nil {
+		return false
+	}
+	// If the number of volumes in the snapshot is not equal to the number of volumes
+	snapshottedVolIDs := vgs.GetVolumeIDs()
+
+	return len(snapshottedVolIDs) == len(volIDs)
+}
+
+func getClusterIDForVolumeID(fms map[string]core.SubVolumeQuiesceClient, volumeID string) string {
+	for _, fm := range fms {
+		for _, vol := range fm.GetVolumes() {
+			if vol.VolumeID == volumeID {
+				return vol.ClusterID
+			}
+		}
+	}
+
+	return ""
+}
+
+func getFsNamesAndSubVolumeFromVolumeIDs(ctx context.Context,
```

Comment:
long functions like this benefit a lot from comments explaining what/how the do it.